### PR TITLE
API: Update `lib.shape_base` namespace

### PR DESCRIPTION
**Pull request by mtsokol**
Relevant issue https://github.com/numpy/numpy/issues/24507

Hi @rgommers @ngoldbaum,

This PR moves `lib.shape_base` module to a private file and ensures that its public methods are only available through the main namespace. There are no additional functions available from a local namespace.

Additionally this PR removes `np.get_array_wrap` from the main namespace as it was already deprecated this month and it isn't used in SciPy, matplotlib, pandas etc. It also deprecates `np.lib.shape_base.get_array_prepare` (copy-paste of `get_array_wrap` implementation) that isn't even exposed in main namespace or used it other libraries.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Despite `get_array_prepare` not being deprecated yet, I think it's probably OK to just go ahead and remove it, since it has even less usage than `get_array_wrap` and it never appeared in the main namespace.

**Comment by rgommers** [comment]
> I think it's probably OK to just go ahead and remove it, since it has even less usage than `get_array_wrap` and it never appeared in the main namespace.

Agreed. It was never public to begin with, and I've never seen it used anywhere.

File: numpy/lib/_shape_base_impl.py [hunk 1]
```diff
@@ -1039,8 +1039,19 @@ def dsplit(ary, indices_or_sections):
...
     """
+
+    # Deprecated in NumPy 2.0, 2023-08-28
+    warnings.warn(
+        "`get_array_prepare` is deprecated. "
+        "(deprecated in NumPy 2.0)",
+        DeprecationWarning,
+        stacklevel=2
+    )
+
```
Comment (ngoldbaum) [inline numpy/lib/_shape_base_impl.py:1054]: IMO no need to deprecate it, just remove it. It has no internal consumers and as far as I can see no public consumers on github besides wrappers for the entire numpy API.

Comment (mtsokol) [reply numpy/lib/_shape_base_impl.py]: I agree that's even better option for that one - removed.

**Comment by ngoldbaum** [comment]
The only change to the main namespace is `get_array_wrap`:

```
In [7]: set(oldall) - set(np.__all__)
Out[7]: {'get_array_wrap'}
```

It also looks like only items available in the main namespace were removed from `np.lib.__all__` (except `get_array_wrap`):

```
In [12]: set(oldliball) - set(np.lib.__all__)
Out[12]: 
{'apply_along_axis',
 'apply_over_axes',
 'array_split',
 'column_stack',
 'dsplit',
 'dstack',
 'expand_dims',
 'get_array_wrap',
 'hsplit',
 'kron',
 'put_along_axis',
 'row_stack',
 'split',
 'take_along_axis',
 'tile',
 'vsplit'}

In [13]: diff = set(oldliball) - set(np.lib.__all__)

In [14]: diff.issubset(np.__all__)
Out[14]: False

In [15]: diff - set(np.__all__)
Out[15]: {'get_array_wrap'}
```

So I think that all the API changes are correct.

**Comment by ngoldbaum** [comment]
Thanks @mtsokol!

**Pull request merged**